AppDB/backup_data.py

```diff
@@ -0,0 +1,367 @@
+    dump_entity.
+
+    Args:
+      entity: The entity to be backed up.
+    Returns:
+      True on success, False otherwise.
+    """
+    key = entity.keys()[0]
+    if re.match(self.PRIVATE_KINDS, key) or re.match(self.PROTECTED_KINDS, key):
+      return False
+    logging.debug("Entity: {0}".format(entity))
+
+    one_entity = entity[key][dbconstants.APP_ENTITY_SCHEMA[0]]
+    if one_entity == datastore_server.TOMBSTONE:
+      return False
+    app_prefix = entity_utils.get_prefix_from_entity_key(key)
+    root_key = entity_utils.get_root_key_from_entity_key(key)
+
+    success = True
+    while True:
+      # Acquire lock.
+      txn_id = self.zoo_keeper.get_transaction_id(app_prefix)
+      try:
+        if self.zoo_keeper.acquire_lock(app_prefix, txn_id, root_key):
+          version = entity[key][dbconstants.APP_ENTITY_SCHEMA[1]]
+          if not self.verify_entity(key, version):
+            # Fetch from the journal.
+            entity = entity_utils.fetch_journal_entry(self.db_access, key)
+            one_entity = entity[key][dbconstants.APP_ENTITY_SCHEMA[0]]
+            if not entity:
+              logging.error("Bad journal entry for key: {0} \nAnd result: {1}".
+                format(key, entity))
+              success = False
+
+          self.dump_entity(one_entity)
+          success = True
+        else:
+          success = False
+      except zk.ZKTransactionException, zk_exception:
+        logging.error("Zookeeper exception {0} while requesting entity lock".
+          format(zk_exception))
+        success = False
+      except zk.ZKInternalException, zk_exception:
+        logging.error("Zookeeper exception {0} while requesting entity lock".
+          format(zk_exception))
+        success = False
+      except dbconstants.AppScaleDBConnectionError, db_exception:
+        logging.error("Database exception {0} while requesting entity lock".
+          format(db_exception))
+        success = False
+      finally:
+        if not success:
+          if not self.zoo_keeper.notify_failed_transaction(app_prefix, txn_id):
+            logging.error("Unable to invalidate txn for {0} with txnid: {1}"\
+              .format(app_prefix, txn_id))
+          logging.error("Failed to backup entity. Retrying shortly...")
+
+        try:
+          self.zoo_keeper.release_lock(app_prefix, txn_id)
+        except zk.ZKTransactionException, zk_exception:
+          logging.error(
+            "Zookeeper exception {0} while releasing entity lock.".
+              format(zk_exception))
+        except zk.ZKInternalException, zk_exception:
+          logging.error(
+            "Zookeeper exception {0} while releasing entity lock.".
+              format(zk_exception))
+
+      if success:
+        break
+      else:
+        time.sleep(self.DB_ERROR_PERIOD)
+
+    return success
+
+  def run_backup(self):
+    """ Runs the backup process. Loops on the entire dataset and dumps it into
+    a file.
+    """
```

Comment:
Forgot the Returns section here. If it only returns True then remove the last return in this function.